### DISCUSSION TITLE
Saving changes to the device (AlienFX-SDK)

### DISCUSSION THREAD

**Discussion by MuaazH** [T-Troll/alienfx-tools#292] [General]
I wrote this code:
``` C++
#include "AlienFX_SDK.h"
#include <Windows.h>

#include <iostream>

using namespace std;

int main(int argc, char **argv) {

    int laptopVid = 0x187c;

    AlienFX_SDK::Afx_action colorRed = {
        AlienFX_SDK::AlienFX_A_Color,                   // type
        100,                                            // time
        100,                                            // tempo
        200,                                            // red
        80,                                             // green
        0,                                             // blue
    };

    AlienFX_SDK::Functions f;
    int pid = f.AlienFXInitialize(laptopVid);
    if (pid < 0) {
        cout << "AlienFXInitialize                    [ FAIL ]" << endl;
        return -1;
    }
    cout << "AlienFXInitialize                    [  OK  ]" << endl;

    while (!f.IsDeviceReady()) {
        cout << "Device not ready" << endl;
        Sleep(50);
    }

    cout << "Detected Version: " << f.GetVersion() << endl;

    if (f.Reset()) {
        cout << "Rest                                 [  OK  ]" << endl;
    } else {
        cout << "Rest                                 [ FAIL ]" << endl;
        return -1;
    }

    if (f.SetColor(0, colorRed)) {
        cout << "SetColor                             [  OK  ]" << endl;
    } else {
        cout << "SetColor                             [ FAIL ]" << endl;
        return -1;
    }

    if (f.UpdateColors()) {
        cout << "UpdateColors                         [  OK  ]" << endl;
    } else {
        cout << "UpdateColors                         [ FAIL ]" << endl;
        return -1;
    }
    while (!f.IsDeviceReady()) {
        cout << "Device not ready" << endl;
        Sleep(50);
    }

    return 0;
}

```

1. This does not work 97% of the time, it only flashes the color. Then it goes back to the last color set by alienfx-tools (alienfx-tools is no longer running btw, I removed it).
2. On the 3% chance that it works, the change is lost when I shutdown the computer.
How do I save the change to the device like alienfx-does?
3. When selecting a color in alienfx-gui the exact thing happends. and the Change only happends when I click save.

I'm not asking because I'm too lazy to read the code. I am reading the code but it's too long & complicated & I'm confused as to why there is a Lights class in the Fans SDK, and Fan controls in alienfx-gui. I'll get to it eventually 😀.

**Comment by T-Troll** [comment]
0. You don't need to check the state or call `Reset()` - it will be done automatically. So you only need to set color and call `UpdateColors()`.
1. Do you have AWCCService running? If yes, it will change colors after every application (including this) quit.
2. Yes, this is different procedure for save colors. Use `SetMultiAction` or `SetPowerAction` for this. BTW, read header (`alienfx_sdk.h`) - all parameters commented there.
3. What gear do you have?

> why there is a Lights class in the Fans SDK

Because some systems have ACPI BIOS lights, handled by Fan SDK at low-level.

> and Fan controls in alienfx-gui

`AlienFX-GUI` is universal, both for lights and fans control.

**Reply by MuaazH** [reply]
Ohhhhhhhhhhh Thank you so much! WORKS GREAT! 🥇

**Comment by MuaazH** [comment]
> 1. This does not work 97% of the time, it only flashes the color.

This could be a bug in the code, I cleaned up the [code](https://github.com/MuaazH/AlienwareControl/tree/main/Lib) and now it works like the official API. I can update the colors for fun, and I can set defaults.

> 2. Do you have AWCCService running?

I don't have AWCCService because I have a messed up version of windows, no modern-apps crap, no windows store, not MSBS (Microsoft Bullshit). AWCC won't even start so I uninstalled it.

I can't really tell where it is, because I changed the code so much. But I think calling `Reset()` in `SetAction` may have something to do with it, since `SetAction` is often called in a loop.

**Reply by T-Troll** [reply]
Ok, let me explain something. APIv4 have 2 different modes:
- Set action for light group (slower)
- Set color for one light (faster)

I check the action type, and set as color if it's single color.
But the issue set color will just blink IN CASE ACTION SET.

For v4, Reset command, in fact, clean and start new action for lights.
Update command store actions.
Reset only called once into light update session until next Update.

So for color set, it reset actions for light to empty.

PS: JFYI - using iostream is so plusish, but adds 200+k to app size immediately (yes, it's SO huge). That's why i prefer good old printf.

**Comment by T-Troll** [comment]
JFLY: I change initialization functions and methods (now much simpler) into latest release.
